Add ChooseTransferAccountPage and allow selecting accounts for wallet transfer

parasharrajat opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!-- Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing -->
$ #3922

### Tests | QA Steps
1. Open Payments page.
2. Click TransferBalance Link.
3. Confirm that the transfer button is disabled if the following conditions are true.
  i. Transfer Amount is less than $ 0.30.
  ii. Which Account is not selected.
4. Click instant Mode. 
5. It should take you to the new page where you will select an account. This page should show you Debit cards only and Add Debit Card Button.
6. Clicking  `+ Add Debit Card` Button opens a new page to add Debit Card.
7. From the Select Account page, click on the any debit card. 
8. You should be taken to the old page where you can see the selected debit card as `Which account?`
9. If your balance is > Fee then `Transfer` Button should be active and it should show the transfer amount - fee. 
4. Click 1-3 Business Days Mode. 
5. It should take you to the new page where you will select an account. This page should show you Bank Accounts only and Add bank account Button.
6. Clicking  `+ Add bank Account` Button opens a new page to add bank Account.
7. From the Select Account page, click on any bank account. 
8. You should be taken to the old page where you can see the selected bank account card as `Which account?`
9. If your balance is > Fee then `Transfer` Button should be active and it should show the transfer amount - fee. 

10. If your balance is > Fee and which account is set.
 i. Click Transfer button.
 ii. Now you should be taken to the Payments Page and a confirm modal should be shown.
 
### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web | Desktop
<!-- Insert screenshots of your changes on the web platform-->

https://user-images.githubusercontent.com/24370807/145723085-ddec27cd-2e8b-4a7c-80e4-76bf04c5a923.mp4

![image](https://user-images.githubusercontent.com/24370807/149260588-f5f706bf-af4a-48cd-ad69-c0aa4d535f4b.png)

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
|||
|----------------|------------|
|![image](https://user-images.githubusercontent.com/24370807/149372389-e7c76b1c-5179-4589-bf33-e337364f681d.png)|![image](https://user-images.githubusercontent.com/24370807/149372454-b4b779e2-df73-4669-b2d9-ec1911a63f7b.png)|

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->
|||
|---------|---------|
|![image](https://user-images.githubusercontent.com/24370807/149370672-ae387b93-a022-4745-a89f-7d378815fb9e.png)|![image](https://user-images.githubusercontent.com/24370807/149370562-501cf1d7-47bb-4ca6-8167-991ca03d37d5.png)|

parasharrajat committed:
Allow hiding the Add payment method button on Payment Method list

parasharrajat committed:
Added ChooseTransferAccontpage

parasharrajat committed:
set navigation between choose account and transfer page

parasharrajat committed:
add methods

parasharrajat committed:
Merge branch 'main' of github.com:Expensify/Expensify.cash into trans-amount-scaffold

parasharrajat committed:
Apply filters correctly

parasharrajat committed:
Reset data on every transfer

parasharrajat committed:
make filtering optional for PaymentMethodList

parasharrajat committed:
UI tweaks

parasharrajat committed:
rename method names

parasharrajat commented:
I didn't do these steps
>  Create a new component that wraps AddPaymentMethodMenu (without modifying this component too much) and call it “AddBalanceTransferAccountButton”

> Create a version of the PaymentMethodList that is selectable. It should not duplicate the logic, but wrap PaymentMethodList.

I used MenuItem directly as there was no need to Popover and disabled the add Payment method button from the paymentMethodList via a prop. There was no new logic needed for the SelectablePaymentMethodList so I used passed the props to filter the list and show the checkboxes.  I don't see any other way of creating a component from `PaymentMethodList ` without passing these props.

parasharrajat committed:
translations fixed

parasharrajat commented:
cc: @marcaaron

marcaaron commented on src/libs/actions/PaymentMethods.js:
```diff
@@ -164,6 +160,20 @@ function saveWalletTransferAmount(transferAmount) {
+/**
+ * @param {String} selectedAccountID
+ */
+function saveWalletTransferAccount(selectedAccountID) {
```
`saveWalletTransferAccountID`

parasharrajat replied:
Thanks.

marcaaron commented on src/libs/actions/PaymentMethods.js:
```diff
@@ -164,6 +160,20 @@ function saveWalletTransferAmount(transferAmount) {
+
+/**
+ * @param {String} filterPaymentMethodType
+ */
```
This doc could use a description because it's not clear what the `filterPaymentMethodType` is used for

>Toggles the user's selected type of payment method (bank account or debit card) on the wallet transfer balance screen.

parasharrajat replied:
Sure

marcaaron commented on src/pages/settings/Payments/ChooseTransferAccountPage.js:
```diff
@@ -1,27 +1,104 @@
+
+const ChooseTransferAccountPage = (props) => {
+    /**
+     * Go back to TransferPage with the selected bank account
```
```suggestion
     * Go back to transfer balance screen with the selected bank account set
```

marcaaron commented on src/pages/settings/Payments/ChooseTransferAccountPage.js:
```diff
@@ -1,27 +1,104 @@
+                        ? props.translate('paymentMethodList.addNewBankAccount')
+                        : props.translate('paymentMethodList.addNewDebitCard')}
+                    icon={Expensicons.Plus}
+                    disabled={props.isLoadingPaymentMethods}
```
What triggers the payment methods to load? Are we able to even access this page if the payment methods are still loading? Should we be able to access it if they are loading?

marcaaron commented on src/pages/settings/Payments/TransferBalancePage.js:
```diff
@@ -119,12 +118,21 @@ class TransferBalancePage extends React.Component {
-    saveTransferAmountAndBalance(transferAmount, selectedAccount) {
+    saveTransferAmountAndStartTransfer(transferAmount, selectedAccount) {
         PaymentMethods.saveWalletTransferAmount(transferAmount);
         PaymentMethods.transferWalletBalance(selectedAccount);
```
NAB but I still do not like that it's difficult to tell which of these is calling the API and which is not.

parasharrajat replied:
I agree with your point here but I will wait for a clear direction from you on handling this. Maybe we can leave this for refactoring later on.

marcaaron replied:
Yes I'm sorry I don't know how to improve it right now and it might not be a problem so "not a blocker"

marcaaron commented on src/pages/settings/Payments/TransferBalancePage.js:
```diff
@@ -119,12 +118,21 @@ class TransferBalancePage extends React.Component {
+    }
+
     render() {
+        this.props.userWallet.currentBalance = 125;
```
❓

parasharrajat replied:
Ops... I was testing it.

parasharrajat committed:
Refactored names, clean up and  comments

parasharrajat commented:
Updated.

parasharrajat committed:
remove unused imports

marcaaron commented on src/libs/actions/PaymentMethods.js:
```diff
@@ -130,7 +130,7 @@ function transferWalletBalance(paymentMethod) {
     Onyx.merge(ONYXKEYS.WALLET_TRANSFER, {loading: true});
 
-    API.TransferWalletBalance(parameters)
+    API.Wallet_TransferBalance(parameters)
```
Instead of changing this can you please fix the export name so it matches the command?

The command name is`TransferWalletBalance` so the `API` method should be `API.TransferWalletBalance()`

parasharrajat replied:
I thought of following the existing code. But I can do that.

marcaaron replied:
Thanks.

stitesExpensify commented on src/pages/settings/Payments/ChooseTransferAccountPage.js:
```diff
@@ -1,27 +1,94 @@
+     * @param {String} paymentType
+     */
+    const navigateToAddPaymentMethodPage = () => {
+        if (props.filterType === CONST.PAYMENT_METHODS.DEBIT_CARD) {
```
Does this prop exist? I don't see it in the propTypes. Is it using the prop from PaymentMethodList?

parasharrajat replied:
Ops, I forgot to refactor it. I will do it asap.

stitesExpensify reviewed (commented):
Testing went pretty well, it's looking good! One bug I found was that when you click add debit card, it always sends you to the add bank account page

parasharrajat committed:
rename and props correction

parasharrajat commented:
Updated @stitesExpensify @marcaaron .

stitesExpensify reviewed (changes requested):
Code looks pretty good to me. I found 2 more bugs while testing though. 

1. If you have 2 accounts and go to the transfer page, then click the ACH button, neither account is selected even though one was selected on the main page
<img width="375" alt="2022-01-14_13-46-18" src="https://user-images.githubusercontent.com/42391420/149583345-0ddc8859-ee11-41cc-82de-27434cefcf34.png">

2. If you have a debit card set as your default, it is not shown on the main transfer page
<img width="373" alt="2022-01-14_14-33-38" src="https://user-images.githubusercontent.com/42391420/149588336-4bbddefa-0bc2-499c-801d-b819e2f88faa.png">

parasharrajat committed:
select default account

parasharrajat commented:
How can I set the default Payment Method?

stitesExpensify commented:
The `userWallet` key in onyx can be manually modified for now to have values like this:
```
{"availableBalance":12345,"bankAccountID":8,"currentBalance":12345,"currentStep":"","tier":1,"tierName":"SILVER","walletLinkedAccountID":1,"walletLinkedAccountType":"debitCard"}
```

You can substitute the `walletLinkedAccountID` and `walletLinkedAccountType` to be whatever you want to test with

parasharrajat committed:
Merge branch 'main' of github.com:Expensify/Expensify.cash into trans-amount-scaffold

parasharrajat committed:
Adopt changes from latest changes

parasharrajat committed:
fix: PaymentMethodlist filtering

parasharrajat committed:
more refactoring

parasharrajat committed:
preselect default account

parasharrajat commented:
Ok. Fixed both issues. 
1. I was never using the default account on Choose Account screen. 
2. I was using the wrong methodID for debit cards. It should be `fundID`. 

Updated. cc: @stitesExpensify

stitesExpensify commented:
I'm still running into some issues while testing. Debit cards seem to work, but if I make a bank account my default the selected account won't change

https://user-images.githubusercontent.com/42391420/150044560-c6dc4146-c691-42b1-9799-73cb71816895.mp4

parasharrajat commented:
Interesting issue. I can't think of the reason why this is happening but I have two accounts now and I will try to test it.

parasharrajat committed:
Merge branch 'main' of github.com:Expensify/Expensify.cash into trans-amount-scaffold

parasharrajat committed:
Fix code to get the selected Account

parasharrajat commented:
Finally, fixed. Thanks for catching these issues. 

I was doing the wrong matching to get the SelectedAccount. It was picking the Defaultaccount if that comes before in the position the selected account in the payment method Array.

Ready for review @marcaaron @stitesExpensify

parasharrajat commented:
This is ready @marcaaron @stitesExpensify . Sorry for the bump.

marcaaron commented:
No worries, but there are [guidelines here](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#timeline-expectations-and-asking-for-help-along-the-way) about when it is appropriate to ping if you need a reminder.

marcaaron commented on src/pages/settings/Payments/TransferBalancePage.js:
```diff
@@ -96,35 +96,62 @@ class TransferBalancePage extends React.Component {
+    }
+
     render() {
         const selectedAccount = this.getSelectedPaymentMethodAccount();
```
e.g. if we are initializing the payment method type in `walletTransfer. accountType` and it is based on the value in `getSelectedPaymentMethodAccount()` then we should be able to just look in `walletTransfer`? Or no?

parasharrajat replied:
When we come back from Choose Transfer Account screen. This helps in picking the selected method. Either we should do that in componentDidUpdate or here.

marcaaron replied:
Yeah I'm not sure that's easy to see or why that would be the case. It maybe works, but the whole interaction between the different sets of data and why we are filtering them on every render is confusing.

marcaaron reviewed (approved):
Approving since the code is acceptable. Left some non blocking comments and ideas for improvements that can potentially be addressed in follow ups.

parasharrajat commented on src/pages/settings/Payments/ChooseTransferAccountPage.js:
```diff
@@ -1,27 +1,100 @@
+                : account.fundID,
+        );
+        Navigation.navigate(ROUTES.SETTINGS_PAYMENTS_TRANSFER_BALANCE);
+    };
```
Yeah, I agree but I didn't see any example of such a functional component. Where should I move this. Outside the component in the same file?

marcaaron replied:
Yeah outside the component in the same file would work. Also maybe a good sign that it should be an action instead of a component method.

parasharrajat commented on src/pages/settings/Payments/ChooseTransferAccountPage.js:
```diff
@@ -1,27 +1,100 @@
+                        shouldShowSelectedState
+                        filterType={props.walletTransfer.filterPaymentMethodType}
+                        selectedMethodID={props.walletTransfer.selectedAccountID}
+                        shouldShowAddPaymentMethodButton={false}
```
Yeah, I was thinking th same but then it would unnecessary refactor for this PR so I left it as it is.

parasharrajat commented on src/pages/settings/Payments/TransferBalancePage.js:
```diff
@@ -96,35 +96,62 @@ class TransferBalancePage extends React.Component {
+        PaymentMethods.saveWalletTransferAccountTypeAndID(
+            selectedAccount.accountType,
+            selectedAccount.methodID,
+        );
```
Yeah that is true. Let me see if I can clean this up bit.

stitesExpensify reviewed (approved):
Code looks good, and it tests well! Nice job 😄

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1747312940) to staging by @stitesExpensify in version: 1.1.32-1 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
